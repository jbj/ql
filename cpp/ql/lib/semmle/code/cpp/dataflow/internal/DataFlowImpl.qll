/**
 * DEPRECATED: Use `semmle.code.cpp.dataflow.new.DataFlow` instead.
 */

import semmle.code.cpp.Location
private import DataFlowImplSpecific
private import codeql.dataflow.internal.DataFlowImpl
import MakeImpl<Location, CppOldDataFlow>
