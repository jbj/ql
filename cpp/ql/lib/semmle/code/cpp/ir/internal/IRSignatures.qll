private import semmle.code.cpp.ir.implementation.IRType

signature module LanguageTypeSig {
  class LanguageType {
    string toString();

    /** Gets a string used in IR dumps */
    string getDumpString();

    /** Gets the size of the type in bytes, if known. */
    int getByteSize();

    /**
     * Holds if the `LanguageType` represents a prvalue of type `type` (if `isGLValue` is `false`), or if
     * it represents a glvalue of type `type` (if `isGLValue` is `true`).
     */
    predicate hasType(OpaqueTypeTag type, boolean isGLValue);

    /**
     * Holds if this type represents the C++ type `type`. If `isGLValue` is `true`, then this type
     * represents a glvalue of type `type`. Otherwise, it represents a prvalue of type `type`.
     */
    predicate hasUnspecifiedType(OpaqueTypeTag type, boolean isGLValue);
  }

  /**
   * The underlying type of types from the language that this IR type is based on.
   */
  class OpaqueTypeTag {
    string toString();
  }

  class TypeDomainType;

  class RealDomainType extends TypeDomainType;

  class ComplexDomainType extends TypeDomainType;

  class ImaginaryDomainType extends TypeDomainType;

  /**
   * Holds if an `IRErrorType` should exist.
   */
  predicate hasErrorType();

  /**
   * Holds if an `IRBooleanType` with the specified `byteSize` should exist.
   */
  predicate hasBooleanType(int byteSize);

  /**
   * Holds if an `IRSignedIntegerType` with the specified `byteSize` should exist.
   */
  predicate hasSignedIntegerType(int byteSize);

  /**
   * Holds if an `IRUnsignedIntegerType` with the specified `byteSize` should exist.
   */
  predicate hasUnsignedIntegerType(int byteSize);

  /**
   * Holds if an `IRFloatingPointType` with the specified size, base, and type domain should exist.
   */
  predicate hasFloatingPointType(int byteSize, int base, TypeDomainType domain);

  /**
   * Holds if an `IRAddressType` with the specified `byteSize` should exist.
   */
  predicate hasAddressType(int byteSize);

  /**
   * Holds if an `IRFunctionAddressType` with the specified `byteSize` should exist.
   */
  predicate hasFunctionAddressType(int byteSize);

  /**
   * Holds if an `IROpaqueType` with the specified `tag` and `byteSize` should exist.
   */
  predicate hasOpaqueType(OpaqueTypeTag tag, int byteSize);

  LanguageType getUnknownType();

  LanguageType getVoidType();

  LanguageType getTypeForPRValue(OpaqueTypeTag type);

  /**
   * Gets the `LanguageType` that represents a prvalue of type `type`, if such a `LanguageType` exists.
   * Otherwise, gets `CppUnknownType`.
   */
  LanguageType getTypeForPRValueOrUnknown(OpaqueTypeTag type);

  /**
   * Gets the `LanguageType` that represents a glvalue of type `type`.
   */
  LanguageType getTypeForGLValue(OpaqueTypeTag type);

  /**
   * Gets the `LanguageType` that represents a prvalue of type `int`.
   */
  LanguageType getIntType();

  /**
   * Gets the `LanguageType` that represents a prvalue of type `bool`.
   */
  LanguageType getBoolType();

  /**
   * Gets the `LanguageType` that represents a glvalue of type `bool`.
   */
  LanguageType getBoolGLValueType();

  /**
   * Gets the `LanguageType` that represents a glvalue of function type.
   */
  LanguageType getFunctionGLValueType();

  /**
   * Gets the `LanguageType` that represents a opaque of unknown type with size `byteSize`.
   */
  LanguageType getUnknownOpaqueType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRBooleanType` with the specified
   * `byteSize`.
   */
  LanguageType getCanonicalBooleanType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRSignedIntegerType` with the specified
   * `byteSize`.
   */
  LanguageType getCanonicalSignedIntegerType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRUnsignedIntegerType` with the specified
   * `byteSize`.
   */
  LanguageType getCanonicalUnsignedIntegerType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRFloatingPointType` with the specified
   * size, base, and type domain.
   */
  LanguageType getCanonicalFloatingPointType(int byteSize, int base, TypeDomainType domain);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRAddressType` with the specified
   * `byteSize`.
   */
  LanguageType getCanonicalAddressType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for an `IRFunctionAddressType` with the specified
   * `byteSize`.
   */
  LanguageType getCanonicalFunctionAddressType(int byteSize);

  /**
   * Gets the `LanguageType` that is the canonical type for `IRErrorType`.
   */
  LanguageType getCanonicalErrorType();

  /**
   * Gets the `LanguageType` that is the canonical type for `IRUnknownType`.
   */
  LanguageType getCanonicalUnknownType();

  /**
   * Gets the `LanguageType` that is the canonical type for `IRVoidType`.
   */
  LanguageType getCanonicalVoidType();

  /**
   * Gets the `LanguageType` that is the canonical type for an `IROpaqueType` with the specified `tag` and
   * `byteSize`.
   */
  LanguageType getCanonicalOpaqueType(OpaqueTypeTag tag, int byteSize);

  /**
   * Gets a string that uniquely identifies an `IROpaqueType` tag. This may be different from the usual
   * `toString()` of the tag in order to ensure uniqueness.
   */
  string getOpaqueTagIdentityString(OpaqueTypeTag tag);
}

signature module LanguageSig {
  class Function;

  class Location;

  class UnknownLocation extends Location;

  class UnknownDefaultLocation extends Location;

  class File;

  class AST;

  class Type;

  class UnknownType extends Type;

  class VoidType extends Type;

  class IntegralType extends Type;

  class FloatingPointType extends Type;

  // REVIEW: May need to synthesize this for other languages. Or do we really need it at all?
  class ClassDerivation;

  class StringLiteral;

  class Variable;

  class AutomaticVariable extends Variable;

  class StaticVariable extends Variable;

  class GlobalVariable extends Variable;

  class Parameter extends Variable;

  class Field extends Variable;

  class BuiltInOperation;

  class Declaration;

  // TODO: Remove necessity for these.
  class Expr;

  class Class; // Used for inheritance conversions

  predicate hasCaseEdge(string minValue, string maxValue);

  predicate hasPositionalArgIndex(int argIndex);

  predicate hasAsmOperandIndex(int operandIndex);

  int getTypeSize(Type type);

  int getPointerSize();

  string getStringLiteralText(StringLiteral s);

  predicate hasPotentialLoop(Function f);

  predicate hasGoto(Function f);

  /**
   * Gets the offset of field `field` in bits.
   */
  int getFieldBitOffset(Field field);

  /**
   * Holds if the specified `Function` can be overridden in a derived class.
   */
  predicate isFunctionVirtual(Function f);
}
