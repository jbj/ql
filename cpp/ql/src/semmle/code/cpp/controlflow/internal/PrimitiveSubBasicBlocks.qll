/**
 * Provides the `PrimitiveSubBasicBlock` class, used for partitioning basic blocks in
 * smaller pieces.
 */
private import PrimitiveBasicBlocks
private import ConstantExprs
private class Node = ControlFlowNodeBase;

/**
 * An abstract class that directs where in the control-flow graph a new
 * `PrimitiveSubBasicBlock` must start. If a `Node` is an instance of this
 * class, that node is guaranteed to be the first node in a `PrimitiveSubBasicBlock`.
 * If multiple libraries use the `PrimitiveSubBasicBlock` library, basic blocks may be
 * split in more places than either library expects, but nothing should break
 * as a direct result of that.
 */
abstract class PrimitiveSubBasicBlockCutNode extends Node {
  PrimitiveSubBasicBlockCutNode() {
    // Some control-flow nodes are not in any basic block. This includes
    // `Conversion`s, expressions that are evaluated at compile time, default
    // arguments, and `Function`s without implementation.
    exists(getPrimitiveBasicBlock(this))
  }
}

/**
 * A block that can be smaller than or equal to a `PrimitiveBasicBlock`. Use this class
 * when `Node` is too fine-grained and `PrimitiveBasicBlock` too
 * coarse-grained. Their successor graph is like that of basic blocks, except
 * that the blocks are split up with an extra edge right before any instance of
 * the abstract class `PrimitiveSubBasicBlockCutNode`. Users of this library must
 * therefore extend `PrimitiveSubBasicBlockCutNode` to direct where basic blocks will be
 * split up.
 */
class PrimitiveSubBasicBlock extends Node {
  PrimitiveSubBasicBlock() {
    this instanceof PrimitiveBasicBlock
    or
    this instanceof PrimitiveSubBasicBlockCutNode
  }

  /** Gets the basic block in which this `PrimitiveSubBasicBlock` is contained. */
  PrimitiveBasicBlock getPrimitiveBasicBlock() {
    result = getPrimitiveBasicBlock(this)
  }

  /**
   * Holds if this `PrimitiveSubBasicBlock` comes first in its basic block. This is the
   * only condition under which a `PrimitiveSubBasicBlock` may have multiple
   * predecessors.
   */
  predicate firstInBB() {
    exists(PrimitiveBasicBlock bb | this.getPosInBasicBlock(bb) = 0)
  }

  /**
   * Holds if this `PrimitiveSubBasicBlock` comes last in its basic block. This is the
   * only condition under which a `PrimitiveSubBasicBlock` may have multiple successors.
   */
  predicate lastInBB() {
    exists(PrimitiveBasicBlock bb |
      this.getPosInBasicBlock(bb) = countPrimitiveSubBasicBlocksInBasicBlock(bb) - 1
    )
  }

  /**
   * Gets the position of this `PrimitiveSubBasicBlock` in its containing basic block
   * `bb`, where `bb` is equal to `getPrimitiveBasicBlock()`.
   */
  int getPosInBasicBlock(PrimitiveBasicBlock bb) {
    exists(int nodePos, int rnk |
      bb = this.getPrimitiveBasicBlock() and
      this = bb.getNode(nodePos) and
      nodePos = rank[rnk](int i | exists(PrimitiveSubBasicBlock n | n = bb.getNode(i))) and
      result = rnk - 1
    )
  }

  /** Gets a successor in the control-flow graph of `PrimitiveSubBasicBlock`s. */
  PrimitiveSubBasicBlock getASuccessor() {
    this.lastInBB() and
    pbbSuccessor(this.getPrimitiveBasicBlock(), result)
    or
    exists(PrimitiveBasicBlock bb |
      result.getPosInBasicBlock(bb) = this.getPosInBasicBlock(bb) + 1
    )
  }

  /**
   * Gets the `pos`th control-flow node in this `PrimitiveSubBasicBlock`. Positions
   * start from 0, and the node at position 0 always exists and compares equal
   * to `this`.
   */
  Node getNode(int pos) {
    exists(PrimitiveBasicBlock bb | bb = this.getPrimitiveBasicBlock() |
      exists(int thisPos | this = bb.getNode(thisPos) |
        result = bb.getNode(thisPos + pos) and
        pos >= 0 and
        pos < this.getNumberOfNodes()
      )
    )
  }

  /** Gets a control-flow node in this `PrimitiveSubBasicBlock`. */
  Node getANode() {
    result = this.getNode(_)
  }

  /** Holds if `this` contains `node`. */
  predicate contains(Node node) {
    node = this.getANode()
  }

  /** Gets a predecessor in the control-flow graph of `PrimitiveSubBasicBlock`s. */
  PrimitiveSubBasicBlock getAPredecessor() {
    result.getASuccessor() = this
  }

  /**
   * Gets the number of control-flow nodes in this `PrimitiveSubBasicBlock`. There is
   * always at least one.
   */
  int getNumberOfNodes() {
    exists(PrimitiveBasicBlock bb | bb = this.getPrimitiveBasicBlock() |
      exists(int thisPos | this = bb.getNode(thisPos) |
        this.lastInBB() and
        result = bb.length() - thisPos
        or
        exists(PrimitiveSubBasicBlock succ, int succPos |
          succ.getPosInBasicBlock(bb) = this.getPosInBasicBlock(bb) + 1 and
          bb.getNode(succPos) = succ and
          result = succPos - thisPos
        )
      )
    )
  }

  /** Gets the last control-flow node in this `PrimitiveSubBasicBlock`. */
  Node getEnd() {
    result = this.getNode(this.getNumberOfNodes() - 1)
  }

  /** Gets the first control-flow node in this `PrimitiveSubBasicBlock`. */
  Node getStart() {
    result = this
  }
}

private predicate pbbSuccessor(PrimitiveBasicBlock b1, PrimitiveBasicBlock b2) {
  b1.getASuccessor() = b2 and
  successors_before_adapted(b1.getNode(b1.length() - 1), b2)
}

private PrimitiveBasicBlock getPrimitiveBasicBlock(Node node) {
  result.getANode() = node
}

/** Gets the number of `PrimitiveSubBasicBlock`s in the given basic block. */
private int countPrimitiveSubBasicBlocksInBasicBlock(PrimitiveBasicBlock bb) {
  result = strictcount(PrimitiveSubBasicBlock sbb | sbb.getPrimitiveBasicBlock() = bb)
}
