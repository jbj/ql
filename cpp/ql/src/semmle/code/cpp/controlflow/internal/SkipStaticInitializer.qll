private import cpp

/**
 * Holds if the expression in this initializer is evaluated at compile time and
 * thus should not have control flow computed.
 */
predicate skipInitializer(Initializer init) {
  exists(LocalVariable local |
    init = local.getInitializer() and
    local.isStatic() and
    not runtimeExprInStaticInitializer(init.getExpr())
  )
}

/**
 * Holds if `e` is an expression in a static initializer that must be evaluated
 * at run time. This predicate computes "is non-const" instead of "is const" in
 * order to avoid recursion through forall.
 */
private predicate runtimeExprInStaticInitializer(Expr e) {
  inStaticInitializer(e) and
  if e instanceof AggregateLiteral
  then runtimeExprInStaticInitializer(e.getAChild())
  else not constantInStaticInitializer(e)
}

private predicate computesConstantAddress(Expr e) {
  addressFromVariableAccess(_, e.getFullyConverted())
  or
  computesConstantAddress(e.getParent())
}

private predicate constantInStaticInitializer(Expr e) {
  inStaticInitializer(e) and
  (
    e.isConstant()
    or
    // This represents the function access as implicitly converted to a pointer
    e instanceof FunctionAccess
    or
    e.(AddressOfExpr).getOperand() instanceof FunctionAccess
    or
    computesConstantAddress(e)
  )
}

/** Holds if `e` is part of the initializer of a local static variable. */
private predicate inStaticInitializer(Expr e) {
  exists(LocalVariable local |
    local.isStatic() and
    e.getParent+() = local.getInitializer()
  )
}

/*
 * The following is adapted from EscapesTree.qll and adjusted to require
 * constants everywhere.
 */

private predicate lvalueToLvalueStep(Expr lvalueIn, Expr lvalueOut) {
  lvalueIn = lvalueOut.(DotFieldAccess).getQualifier().getFullyConverted()
  or
  lvalueIn.getConversion() = lvalueOut.(ParenthesisExpr)
  or
  // When an object is implicitly converted to a reference to one of its base
  // classes, it gets two `Conversion`s: there is first an implicit
  // `CStyleCast` to its base class followed by a `ReferenceToExpr` to a
  // reference to its base class. Whereas an explicit cast to the base class
  // would produce an rvalue, which would not be convertible to an lvalue
  // reference, this implicit cast instead produces an lvalue. The following
  // case ensures that we propagate the property of being an lvalue through
  // such casts.
  lvalueIn.getConversion() = lvalueOut and
  lvalueOut.(CStyleCast).isImplicit()
}

private predicate pointerToLvalueStep(Expr pointerIn, Expr lvalueOut) {
  lvalueOut = any(ArrayExpr ae |
    pointerIn = ae.getArrayBase().getFullyConverted() and
    ae.getArrayOffset().isConstant()
  )
  or
  pointerIn = lvalueOut.(PointerDereferenceExpr).getOperand().getFullyConverted()
  or
  pointerIn = lvalueOut.(PointerFieldAccess).getQualifier().getFullyConverted()
}

private predicate lvalueToPointerStep(Expr lvalueIn, Expr pointerOut) {
  lvalueIn.getConversion() = pointerOut.(ArrayToPointerConversion)
  or
  lvalueIn = pointerOut.(AddressOfExpr).getOperand().getFullyConverted()
}

private predicate pointerToPointerStep(Expr pointerIn, Expr pointerOut) {
  (
    pointerOut instanceof PointerAddExpr
    or
    pointerOut instanceof PointerSubExpr
  ) and
  pointerIn = pointerOut.getAChild().getFullyConverted() and
  pointerIn.getType().getUnspecifiedType() instanceof PointerType and
  // The pointer arg won't be constant in the sense of `isConstant`, so this
  // will have to match the integer argument.
  pointerOut.getAChild().isConstant()
  or
  pointerIn = pointerOut.(UnaryPlusExpr).getOperand().getFullyConverted()
  or
  pointerIn.getConversion() = pointerOut.(Cast)
  or
  pointerIn.getConversion() = pointerOut.(ParenthesisExpr)
  or
  pointerOut = any(ConditionalExpr cond |
    cond.getCondition().getValue().toInt() != 0 and
    pointerIn = cond.getThen().getFullyConverted()
    or
    cond.getCondition().getValue().toInt() = 0 and
    pointerIn = cond.getElse().getFullyConverted()
  )
}

private predicate lvalueToReferenceStep(Expr lvalueIn, Expr referenceOut) {
  lvalueIn.getConversion() = referenceOut.(ReferenceToExpr)
}

private predicate referenceToLvalueStep(Expr referenceIn, Expr lvalueOut) {
  // This probably cannot happen. It would require an expression to be
  // converted to a reference and back again without an intermediate variable
  // assignment.
  referenceIn.getConversion() = lvalueOut.(ReferenceDereferenceExpr)
}

private predicate referenceToReferenceStep(Expr referenceIn, Expr referenceOut) {
  referenceIn.getConversion() = referenceOut.(Cast)
  or
  referenceIn.getConversion() = referenceOut.(ParenthesisExpr)
}

private predicate lvalueFromVariableAccess(VariableAccess va, Expr lvalue) {
  // Base case for non-reference types.
  lvalue = va and
  not va.getConversion() instanceof ReferenceDereferenceExpr and
  inStaticInitializer(va) and
  va.getTarget() = any(Variable v |
    v.isStatic()
    or
    v instanceof GlobalOrNamespaceVariable
  )
  or
  // Base case for reference types where we pretend that they are
  // non-reference types. The type of the target of `va` can be `ReferenceType`
  // or `FunctionReferenceType`.
  lvalue = va.getConversion().(ReferenceDereferenceExpr)
  or
  // lvalue -> lvalue
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToLvalueStep(prev, lvalue)
  )
  or
  // pointer -> lvalue
  exists(Expr prev |
    pointerFromVariableAccess(va, prev) and
    pointerToLvalueStep(prev, lvalue)
  )
  or
  // reference -> lvalue
  exists(Expr prev |
    referenceFromVariableAccess(va, prev) and
    referenceToLvalueStep(prev, lvalue)
  )
}

private predicate pointerFromVariableAccess(VariableAccess va, Expr pointer) {
  // pointer -> pointer
  exists(Expr prev |
    pointerFromVariableAccess(va, prev) and
    pointerToPointerStep(prev, pointer)
  )
  or
  // lvalue -> pointer
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToPointerStep(prev, pointer)
  )
}

private predicate referenceFromVariableAccess(VariableAccess va, Expr reference) {
  // reference -> reference
  exists(Expr prev |
    referenceFromVariableAccess(va, prev) and
    referenceToReferenceStep(prev, reference)
  )
  or
  // lvalue -> reference
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToReferenceStep(prev, reference)
  )
}

/**
 * Holds if `e` is a fully-converted expression that evaluates to an address
 * derived from the address of `va`.
 */
private predicate addressFromVariableAccess(VariableAccess va, Expr e) {
  pointerFromVariableAccess(va, e)
  or
  referenceFromVariableAccess(va, e)
  or
  // `e` could be a pointer that is converted to a reference as the final step,
  // meaning that we pass a value that is two dereferences away from referring
  // to `va`. This happens, for example, with `void std::vector::push_back(T&&
  // value);` when called as `v.push_back(&x)`, for a static variable `x`. It
  // can also happen when taking a reference to a const pointer to a
  // (potentially non-const) value.
  exists(Expr pointerValue |
    pointerFromVariableAccess(va, pointerValue) and
    e = pointerValue.getConversion().(ReferenceToExpr)
  )
}
