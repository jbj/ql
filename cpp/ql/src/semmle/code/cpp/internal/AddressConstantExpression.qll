private import cpp

cached
predicate addressConstantExpression(Expr e) {
// TODO: propagate these underscores as far down as possible.
  pointerFromVariableAccess(_, e)
  or
  referenceFromVariableAccess(_, e)
  or
  // `e` could be a pointer that is converted to a reference as the final step,
  // meaning that we pass a value that is two dereferences away from referring
  // to `va`. This happens, for example, with `void std::vector::push_back(T&&
  // value);` when called as `v.push_back(&x)`, for a static variable `x`. It
  // can also happen when taking a reference to a const pointer to a
  // (potentially non-const) value.
  exists(Expr pointerValue |
    pointerFromVariableAccess(_, pointerValue) and
    e = pointerValue.getConversion().(ReferenceToExpr)
  )
  or
  isFunctionAccess(e)
}

/**
 * Holds if `e` is a constant function address. Implicit conversions mean that
 * an expression like `myFunction` is equivalent to `&**&***myFunction`.
 */
private predicate isFunctionAccess(Expr e) {
  e instanceof FunctionAccess
  or
  isFunctionAccess(e.(AddressOfExpr).getOperand())
  or
  isFunctionAccess(e.(PointerDereferenceExpr).getOperand())
  // TODO: what about conversions? What about C++ function _references_?
}

/*
 * The following is adapted from EscapesTree.qll and adjusted to require
 * constants everywhere.
 */

private predicate lvalueToLvalueStep(Expr lvalueIn, Expr lvalueOut) {
  lvalueIn = lvalueOut.(DotFieldAccess).getQualifier().getFullyConverted()
  or
  lvalueIn.getConversion() = lvalueOut.(ParenthesisExpr)
}

private predicate pointerToLvalueStep(Expr pointerIn, Expr lvalueOut) {
  lvalueOut = any(ArrayExpr ae |
    pointerIn = ae.getArrayBase().getFullyConverted() and
    ae.getArrayOffset().isConstant()
  )
  or
  pointerIn = lvalueOut.(PointerDereferenceExpr).getOperand().getFullyConverted()
  or
  pointerIn = lvalueOut.(PointerFieldAccess).getQualifier().getFullyConverted()
}

private predicate lvalueToPointerStep(Expr lvalueIn, Expr pointerOut) {
  lvalueIn.getConversion() = pointerOut.(ArrayToPointerConversion)
  or
  lvalueIn = pointerOut.(AddressOfExpr).getOperand().getFullyConverted()
}

private predicate pointerToPointerStep(Expr pointerIn, Expr pointerOut) {
  (
    pointerOut instanceof PointerAddExpr
    or
    pointerOut instanceof PointerSubExpr
  ) and
  pointerIn = pointerOut.getAChild().getFullyConverted() and
  pointerIn.getType().getUnspecifiedType() instanceof PointerType and
  // The pointer arg won't be constant in the sense of `isConstant`, so this
  // will have to match the integer argument.
  pointerOut.getAChild().isConstant()
  or
  pointerIn = pointerOut.(UnaryPlusExpr).getOperand().getFullyConverted()
  or
  pointerIn.getConversion() = pointerOut.(Cast)
  or
  pointerIn.getConversion() = pointerOut.(ParenthesisExpr)
  or
  pointerOut = any(ConditionalExpr cond |
    cond.getCondition().getValue().toInt() != 0 and
    pointerIn = cond.getThen().getFullyConverted()
    or
    cond.getCondition().getValue().toInt() = 0 and
    pointerIn = cond.getElse().getFullyConverted()
  )
}

private predicate lvalueToReferenceStep(Expr lvalueIn, Expr referenceOut) {
  lvalueIn.getConversion() = referenceOut.(ReferenceToExpr)
}

private predicate referenceToLvalueStep(Expr referenceIn, Expr lvalueOut) {
  // This probably cannot happen. It would require an expression to be
  // converted to a reference and back again without an intermediate variable
  // assignment.
  referenceIn.getConversion() = lvalueOut.(ReferenceDereferenceExpr)
}

private predicate referenceToReferenceStep(Expr referenceIn, Expr referenceOut) {
  referenceIn.getConversion() = referenceOut.(Cast)
  or
  referenceIn.getConversion() = referenceOut.(ParenthesisExpr)
}

private predicate lvalueFromVariableAccess(VariableAccess va, Expr lvalue) {
  // Base case for non-reference types.
  lvalue = va and
  not va.getConversion() instanceof ReferenceDereferenceExpr and
  va.getTarget() = any(Variable v |
    v.isStatic()
    or
    v instanceof GlobalOrNamespaceVariable
  )
  or
  // Base case for reference types where we pretend that they are
  // non-reference types. The type of the target of `va` can be `ReferenceType`
  // or `FunctionReferenceType`.
  lvalue = va.getConversion().(ReferenceDereferenceExpr)
  or
  // lvalue -> lvalue
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToLvalueStep(prev, lvalue)
  )
  or
  // pointer -> lvalue
  exists(Expr prev |
    pointerFromVariableAccess(va, prev) and
    pointerToLvalueStep(prev, lvalue)
  )
  or
  // reference -> lvalue
  exists(Expr prev |
    referenceFromVariableAccess(va, prev) and
    referenceToLvalueStep(prev, lvalue)
  )
}

private predicate pointerFromVariableAccess(VariableAccess va, Expr pointer) {
  // pointer -> pointer
  exists(Expr prev |
    pointerFromVariableAccess(va, prev) and
    pointerToPointerStep(prev, pointer)
  )
  or
  // lvalue -> pointer
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToPointerStep(prev, pointer)
  )
}

private predicate referenceFromVariableAccess(VariableAccess va, Expr reference) {
  // reference -> reference
  exists(Expr prev |
    referenceFromVariableAccess(va, prev) and
    referenceToReferenceStep(prev, reference)
  )
  or
  // lvalue -> reference
  exists(Expr prev |
    lvalueFromVariableAccess(va, prev) and
    lvalueToReferenceStep(prev, reference)
  )
}
