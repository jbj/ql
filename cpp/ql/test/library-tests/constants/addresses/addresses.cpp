// semmle-extractor-options: --c++14

const int int_const = 1;
extern const int extern_int_const = 1;
extern const int extern_int_const_noinit;

int int_var;
int int_arr[4];
int int_arr_arr[4][4];

void sideEffect();
using fptr = void (*)();
using fref = void (&)();



// All variables in this function are initialized to constants, as witnessed by
// the `constexpr` annotation that compiles under C++14.
void constantAddresses(int param) {
    constexpr int *p1 = &int_var;
    constexpr int *p2 = &*&int_var;
    constexpr int *p3 = &int_arr[1] + 1;
    constexpr int (*p4)[4] = &int_arr_arr[1] + 1;
    constexpr int *p5 = &int_arr_arr[1][1] + 1;
    constexpr int *p6 = &int_arr_arr[int_const][extern_int_const];

    // Commented out because clang and EDG disagree on whether this is
    // constant.
    //constexpr int *p7 = &int_arr[ ({ 1; }) ]; // StmtExpr

    constexpr int *p8 = &int_arr[ ((void)0, 1) ]; // CommaExpr
    constexpr int *p9 = &int_arr[int_const ? 1 : param];
    constexpr int *p10 = &int_arr[&int_arr + 1 ? 1 : param]; // recursive
    constexpr int *p11 = &int_arr[&int_arr[1] - &int_arr[0]];

    constexpr int *p12 = p11 + 1;

    constexpr int &r1 = int_var;
    constexpr int &r2 = int_arr_arr[1][1];
    constexpr int &r4 = r2;
    constexpr int &r5 = *(&r2 + 1);

    constexpr fptr fp1 = sideEffect;
    constexpr fptr fp2 = &sideEffect;
    constexpr fptr fp3 = **&**&**sideEffect;

    constexpr fref fr1 = sideEffect;
    constexpr fref fr2 = *&sideEffect;
    constexpr fref fr3 = **sideEffect;
    constexpr fref fr4 = fr1;
}




// All variables in this function are initialized to non-const values. Writing
// `const` in front of any of the variables will be a compile error (C++14).
void nonConstantAddresses(int param, int *pparam, int &rparam, fref frparam) {
    int *p3 = &int_arr[param];
    int *p6 = &int_arr[extern_int_const_noinit];

    int *p7 = &int_arr[ ({ sideEffect(); 1; }) ]; // StmtExpr
    int *p8 = &int_arr[ (sideEffect(), 1) ]; // CommaExpr
    int *p9 = &int_arr[int_const ? param : 1];
    int *p10 = &int_arr[&int_arr + 1 ? param : 1]; // recursive
    int *p12 = pparam;

    int *&&r3 = &int_var; // reference to temporary is not a const
    int &r4 = rparam;

    fref fr4 = frparam;
}
