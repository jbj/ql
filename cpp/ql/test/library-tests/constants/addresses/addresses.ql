import cpp
import semmle.code.cpp.internal.AddressConstantExpression

from Expr e, string msg, LocalVariable var, Function f
where
  e = var.getInitializer().getExpr().getFullyConverted() and
  var.getFunction() = f and
  (
    addressConstantExpression(e) and
    f.getName() = "nonConstantAddresses" and
    msg = "misclassified as constant"
    or
    not addressConstantExpression(e) and
    f.getName() = "constantAddresses" and
    msg = "misclassified as NOT constant"
  )
select e, var.getName(), msg
