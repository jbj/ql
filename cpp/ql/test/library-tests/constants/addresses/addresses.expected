| addresses.cpp:36:26:36:32 | ... + ... | p12 | misclassified as NOT constant |
| addresses.cpp:71:15:71:20 | (reference to) | r4 | misclassified as constant |
| addresses.cpp:73:16:73:22 | (reference to) | fr4 | misclassified as constant |
